### PR TITLE
Add documentation to call async python from rust

### PR DESCRIPTION
**Pull request by sansyrox**
Thank you for contributing to pyo3!

Please consider adding the following to your pull request:
 - an entry in CHANGELOG.md
 - docs to all new functions and / or detail in the guide
 - tests for all new or changed functions

Be aware the CI pipeline will check your pull request for the following:
 - Rust tests (Just `cargo test` or `make test` if you need to test examples)
 - Rust lints (`make clippy`)
 - Rust formatting (`cargo fmt`)
 - Python formatting (`black . --check`. You can install black with `pip install black`)
 - Compatibility with all supported Python versions for all examples. This uses `tox`; you can do run it using `make test_py`.

You can run a similar set of checks as the CI pipeline using `make test`.

## Unable to run the docs book
I am running the book using the following command and am getting an error:

```
npx mdbook build -d ../gh-pages-build guide --open
npm ERR! could not determine executable to run

npm ERR! A complete log of this run can be found in:
Add documentation to call async python from rust
npm ERR!     /Users/bruhh/.npm/_logs/2021-06-13T20_04_12_108Z-debug.log
```

How do I run the book to check if everything is working well?

### CODE REVIEW

File: guide/src/async_python_from_rust.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+```
+
+Above, we have defined an async function. How we send it to Rust code will make a difference.
+
+In python, even an async function is treated like a regular function until called.
+e.g.
+
+```python
+print(type fx) # this will output the type function
+print(type fx()) # this will output the type of a coroutine
```
Comment (davidhewitt) [inline guide/src/async_python_from_rust.md:21]: I prefer to use REPL syntax for these kinds of examples:

```suggestion
>>> type(fx)
<class 'function'>
>>> type(fx())
<class 'coroutine'>
```

File: guide/src/async_python_from_rust.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+## Async Functions
+
+Before we start, let us take a look at the following snippet.
+
+```python
+async def fx():
+    return "hello, world"
+```
+
+Above, we have defined an async function. How we send it to Rust code will make a difference.
```
Comment (davidhewitt) [inline guide/src/async_python_from_rust.md:14]: > How we send it to Rust code will make a difference.

Can you try extend this sentence? It's not very clear what you mean here. The reader might wonder why does it make a difference? How can they send it to Rust code?

File: guide/src/async_python_from_rust.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
+# Calling async Python functions in Rust code
+
+In the previous chapter, we learnt how to call Python functions in Rust code. Now, we will look at two ways to call Python async functions from Rust code.
```
Comment (davidhewitt) [inline guide/src/async_python_from_rust.md:3]: > Now, we will look at two ways to call Python async functions from Rust code.

Can we speak briefly about each of these two ways in this introduction?

File: guide/src/async_python_from_rust.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            // converting into a Future is mandatory to be awaited in rust
+            let f = pyo3_asyncio::into_future(&coro).unwrap()
+
+            pyo3_asyncio::async_std::run_until_complete(py, async move {
+                f.await?; // here the async function is called
+                Ok(())
+            });
+        });
+        // now we will need to convert the function to a coroutine
+    }
```
Comment (davidhewitt) [inline guide/src/async_python_from_rust.md:117]: No need to use `with_gil` here, can take `Python` as an argument:

```suggestion
    fn start(&self, py: Python, python_function: &PyAny) {
        // here we are converting the function to a coroutine by calling it
        let coro = handler.as_ref(py).call0().unwrap();
        // here we are converting the coroutine to a Future
        // converting into a Future is mandatory to be awaited in rust
        let f = pyo3_asyncio::into_future(&coro).unwrap()

        pyo3_asyncio::async_std::run_until_complete(py, async move {
            f.await?; // here the async function is called
            Ok(())
        });
        // now we will need to convert the function to a coroutine
    }
```

File: guide/src/async_python_from_rust.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
```
Comment (davidhewitt) [inline guide/src/async_python_from_rust.md:164]:
```suggestion
    fn start(&self, py: Python, python_function: &PyAny) {
        // where python_function is the function added from python
        // here we are converting the coroutine to a Future
        // converting into a Future is mandatory to be awaited in rust
        let f = pyo3_asyncio::into_future(&coro).unwrap()

        pyo3_asyncio::async_std::run_until_complete(py, async move {
            f.await?; // here the async function is called
            Ok(())
        });
        // now we will need to convert the function to a coroutine
    }
```

File: guide/src/async_python_from_rust.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+In python, even an async function is treated like a regular function until called.
+e.g.
+
+```python
+print(type fx) # this will output the type function
+print(type fx()) # this will output the type of a coroutine
+```
+
+The native object `PyAny` can be used to call the Python functions in Rust code.
```
Comment (davidhewitt) [inline guide/src/async_python_from_rust.md:24]:
```suggestion
The native object `PyAny` can be used to call Python functions in Rust code.
```

File: guide/src/async_python_from_rust.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+Now, let us take a look at how will we share the Python function in our rust code.
```
Comment (davidhewitt) [inline guide/src/async_python_from_rust.md:26]: Can we add some more detail here about what we will do, so that the reader can feel a sense of direction?

File: guide/src/async_python_from_rust.md [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+
+The `start` function shows how we will be able to send our Python code to rust.
+
+Now, we have two choices. Either we can pass the Python function as the following:
+
+```python
+some_object.start(fx)
+```
+
+or, we can pass an awaitable coroutine
```
Comment (davidhewitt) [inline guide/src/async_python_from_rust.md:78]:
```suggestion
or we can pass an awaitable coroutine:
```

**Review by davidhewitt — COMMENTED** [review]
Thanks for adding this. I think a few edits will be needed to make it flow a bit easier. I've made a couple of suggestions, we might need to re-read again once those are made.

Also, can you please move the contents of this to `async-await.md`? I don't think we need a new chapter just for this part.

> How do I run the book to check if everything is working well?

I think the following should do it:

```bash
cargo install mdbook
mdbook build -d ../gh-pages-build guide --open
```

**Comment by sansyrox** [comment]
> Thanks for adding this. I think a few edits will be needed to make it flow a bit easier. I've made a couple of suggestions, we might need to re-read again once those are made.
> 
> Also, can you please move the contents of this to `async-await.md`? I don't think we need a new chapter just for this part.
> 
> > How do I run the book to check if everything is working well?
> 
> I think the following should do it:
> 
> ```shell
> cargo install mdbook
> mdbook build -d ../gh-pages-build guide --open
> ```

These changes make sense @davidhewitt . I will make changes on the weekend 😄

**Comment by sansyrox** [comment]
@davidhewitt , I have made some updates. Please let me know if this works.

File: guide/src/ecosystem/async-await.md [hunk 9]
```diff
@@ -268,3 +269,191 @@ async fn main() -> pyo3::PyResult<()> {
...
+
+In python, even an async function is treated like a regular function until called.
+e.g.
+
+```python
+>>> type(fx)
+<class 'function'>
+>>> type(fx())
+<class 'coroutine'>
+```
```
Comment (davidhewitt) [inline guide/src/ecosystem/async-await.md:300]: The start of the `async / await` chapter already describes what an `async` Python is. I dont' think we need to repeat this, and can instead replace this with a summary / recap:

```suggestion
Above we learned that an async Python function is a special category of function which returns a `coroutine` object. We learned how to write these functions in Rust, so that they could be awaited in Python.
```

Your next sentence "we will now write Rust functions" is then a perfect continuation of the chapter.

File: guide/src/python_from_rust.md [hunk 10]
```diff
@@ -1,22 +1,23 @@
 # Calling Python in Rust code
```
Comment (davidhewitt) [inline guide/src/python_from_rust.md:1]: Please revert the changes to this file... look like formatting changes?

File: guide/src/ecosystem/async-await.md [hunk 11]
```diff
@@ -268,3 +269,191 @@ async fn main() -> pyo3::PyResult<()> {
...
+
+We will now write rust functions to execute the python code in rust. The `PyAny` binding will be used to call Python functions in Rust code.
```
Comment (davidhewitt) [inline guide/src/ecosystem/async-await.md:302]:
```suggestion
We will now write Rust functions which can await Python coroutines. The `PyAny` binding will be used to pass `coroutine` arguments. Again the `pyo3_asyncio` crate wll help us complete the implementation.
```

File: guide/src/ecosystem/async-await.md [hunk 12]
```diff
@@ -268,3 +269,191 @@ async fn main() -> pyo3::PyResult<()> {
...
+some_object.start(fx)
+```
+
+or we can pass an awaitable coroutine
+
+```python
+some_object.start(fx())
+```
+
+In both cases, we will have to take the help of an external crate called [pyo3_asyncio](https://crates.io/crates/pyo3-asyncio)
```
Comment (davidhewitt) [inline guide/src/ecosystem/async-await.md:362]: No need to introduce this here, as it's already used throughout the chapter.

```suggestion
```

File: guide/src/ecosystem/async-await.md [hunk 13]
```diff
@@ -268,3 +269,191 @@ async fn main() -> pyo3::PyResult<()> {
...
+#[pymodule]
+pub fn some_module(py: Python<'_>, m: &PyModule) -> PyResult<()> {
+    m.add_class::<Server>()?;
+    pyo3_asyncio::try_init(py);
+    Ok(())
+}
+
+```
+
+Now, a _major_ reason where you would want to choose one approach over another is that a pre made coroutine cannot be awaited more than once.
```
Comment (davidhewitt) [inline guide/src/ecosystem/async-await.md:448]: Make a new sub-section "reusing a coroutine", perhaps?

```suggestion
## Reusing a coroutine
```

File: guide/src/ecosystem/async-await.md [hunk 14]
```diff
@@ -268,3 +269,191 @@ async fn main() -> pyo3::PyResult<()> {
...
+
+```python
+some_object.start(fx)
+```
+
+or we can pass an awaitable coroutine
+
+```python
+some_object.start(fx())
+```
```
Comment (davidhewitt) [inline guide/src/ecosystem/async-await.md:360]: I think that introducing this choice already is confusing. Let's just focus on how to pass an awaitable coroutine for now, and use a new section at the end to talk about passing the function (I've suggested calling it "reusing a coroutine").

```suggestion
The following Python code shows how to pass a Python coroutine to the above Rust code:

```python
import some_module

async def fx():
    return "hello world"

some_object.await_in_rust(fx())
```
```

File: guide/src/ecosystem/async-await.md [hunk 15]
```diff
@@ -268,3 +269,191 @@ async fn main() -> pyo3::PyResult<()> {
...
+        // where python_function is the function added from python
+    }
+}
+
+#[pymodule]
+pub fn some_module(py: Python<'_>, m: &PyModule) -> PyResult<()> {
+    m.add_class::<Server>()?;
+    pyo3_asyncio::try_init(py);
+    Ok(())
+}
```
Comment (davidhewitt) [inline guide/src/ecosystem/async-await.md:330]: I think that using `#[pyclass]` is too complex for this example, let's use a `#[pyfunction]` instead. It will make it easier for the user to focus on the relevant code.

All the examples below will also need to be changed.

```suggestion
use pyo3::prelude::*;

/// Awaits a Python coroutine from Rust code.
fn await_in_rust(coroutine: &PyAny) {
    // we will fill in this implementation as the section continues...
}

// We need to create a module to contain this function
#[pymodule]
pub fn some_module(py: Python<'_>, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(await_in_rust, m))?;
    pyo3_asyncio::try_init(py);
    Ok(())
}
```

File: guide/src/ecosystem/async-await.md [hunk 16]
```diff
@@ -268,3 +269,191 @@ async fn main() -> pyo3::PyResult<()> {
...
+
+```
+
+Now, if we had passed a pre-made python coroutine , we wouldn't have to write the following code snippet:
+
+```rust
+let coro = handler.as_ref(py).call0().unwrap();
+```
```
Comment (davidhewitt) [inline guide/src/ecosystem/async-await.md:408]: As discussed, let's move this to the end and instead show the user that they need to add this line to execute the async function.

File: guide/src/ecosystem/async-await.md [hunk 17]
```diff
@@ -268,3 +269,191 @@ async fn main() -> pyo3::PyResult<()> {
...
+
+e.g. if you are planning to store one coroutine and call it multiple times for some reason, it won't be possible and you would be getting the error
+
+```rust
+RuntimeError: cannot reuse already awaited coroutine
+```
+
+So, if you are planning to store one function and reuse it again. It is sensible to pass it as a function, otherwise you can also pass it as a coroutine.
+
+You can read more about the crate [here](https://crates.io/crates/pyo3-asyncio).
```
Comment (davidhewitt) [inline guide/src/ecosystem/async-await.md:458]: No need for this, covered a lot in this chapter already.

```suggestion
```

**Comment by davidhewitt** [comment]
@sansyrox thanks again for the changes, and sorry it has taken me so long to review.

I think the content is good but we still need to make some adjustments to make it easier for the reader to follow. I've made some suggestions which I think introduce the ideas to the reader more gradually as well as telling them in advance what they can expect to learn.

**Comment by sansyrox** [comment]
@davidhewitt , sorry for my late revert. I had a busy week. I will make the changes suggested by you! 🖖

**Comment by sansyrox** [comment]
@davidhewitt , I have updated the docs. Let me know what you think now?

**Comment by davidhewitt** [comment]
Thanks, looks fine to me now! There were some other changes to this guide in the meantime, so I've rebased to merge the changes and will merge this now.
